examples/NeuralNetwork/neural_network.dx

```diff
@@ -0,0 +1,211 @@
+    comp x = 1.0 - x
+    fcrossent f = f(tar) * log(out + eps)
+    (fcrossent id) + (fcrossent comp)
+
+binaryCrossEntropy [1.0] [0.99]
+
+binaryCrossEntropy [1.0] [0.01]
+
+binaryCrossEntropy [0.01] [1.0]
+
+
+' ### Layers
+The type of a Affine layer is: `(d=>n=>Real, d=>Real)`.
+That is the weight matrix and the bias vector.
+
+initAffine:: A n::Ix d::Ix. Key->(d=>n=>Real, d=>Real)
+initAffine key = ..
+    (weightKey, biasKey) = splitKey key
+    weights = for x::d. for y::n.
+        0.1 * (randn $ ixkey2 weightKey x y)
+    biases = for y::d.
+        0.1 * (randn $ ixkey biasKey y)
+    (weights, biases)
+
+initAffine @3 @2 (newKey 0)
+> ( [ [-1.6417312e-2, -0.11909912]
+> , [-1.777558e-2, 8.824792e-2]
+> , [0.12012412, 0.12777644] ]
+> , [0.12112768, 2.3284968e-2] )
+
+applyAffine:: (d=>n=>Real, d=>Real) -> (n=>Real) -> (d=>Real)
+applyAffine (w, b) z = for k::d.
+    (mvp w z).k + b.k
+
+
+' try it out
+
+layer1 = initAffine @3 @2 (newKey 1)
+input1 = [1.0, 2.0, 3.0]
+sigmoid $ applyAffine layer1 input1
+> [2.863232e-2, 0.22663473]
+
+' ## XOR Task
+
+xor:: Bool -> Bool -> Bool
+xor a b = (a && (not b)) || ((not a) && b)
+
+xor True True
+> False
+
+xor False False
+> False
+
+xor True False
+> True
+
+xor False True
+> True
+
+trueXorClass:: (2=>Real)->Bool
+trueXorClass x = xor (x.(0@2) > 0.5) (x.(1@2) > 0.5)
+
+trueXorClass [0.3, 0.3]
+> False
+
+trueXorClass [0.3, 0.7]
+> True
+
+trueXorClass [0.7, 0.3]
+> True
+
+trueXorClass [0.7, 0.7]
+> False
+
+xorTrainInputs = for i::10. for j::2.
+    trainKey = newKey 42
+    rand (ixkey2 trainKey i j)
+
+xorTrainTargetClasses = map trueXorClass xorTrainInputs
+xorTrainTargets = for i. for j::1.
+    b2r xorTrainTargetClasses.i
+
+
+xorTrainInputs
+
+xorTrainTargetClasses
+
+xorTrainTargets
+
+(filter trueXorClass xorTrainInputs).(asidx 1)
+
+' ## Xor Net
+This is a two layer net.
+AFAICT no data structure in Dex right now can hold a arbitrary collection
+of neural network layers/parameters, as it would be a heterogeneous vector, or a tuple of non-fixed length
+As such need different functions for each depth of network
+
+xorNetParams = ( ..
+    initAffine @2 @6 (newKey 10), ..
+    initAffine @6 @1 (newKey 20) ..
+)
+
+:t xorNetParams
+
+apply2Net:: ((hz=>(iz=>Real), hz=>Real), (oz=>(hz=>Real), oz=>Real)) -> iz=>Real -> oz=>Real
+apply2Net (layer1, layer2) input = 
+    sigmoid $ applyAffine layer2 $ sigmoid $ applyAffine layer1 input
+
+apply2Net xorNetParams [0.9, 0.1]
+
+(binaryCrossEntropy [1.0]) $ apply2Net xorNetParams [0.9, 0.1]
+
+eval2NetLoss:: (oz=>Real->oz=>Real->Real) -> oz=>Real -> iz=>Real -> ((hz=>(iz=>Real), hz=>Real), (oz=>(hz=>Real), oz=>Real)) -> Real
+eval2NetLoss loss target input netParams = ..
+    loss target $ apply2Net netParams input
+
+eval2NetLoss binaryCrossEntropy [1.0] [0.9, 0.1] xorNetParams 
+
+xorLoss:: ((hz=>(2=>Real), hz=>Real), (1=>(hz=>Real), 1=>Real)) -> Real
+xorLoss = eval2NetLoss binaryCrossEntropy [1.0] [0.9, 0.1]
+
+
+grad xorLoss xorNetParams
```

Comment:
this currently errors:


```
Zeros not implemented for Int
CallStack (from HasCallStack):
  error, called at src/lib/Embed.hs:189:13 in dex-0.1.0.0-9pQ6aTLha3fLdCjkbXPWRy:Embed
```

Reply:
I am hoping `grad` on a **Tuple of Tuples of Arrays**, works and returns  a **Tuple of Tuples of Arrays** of derivatives.
Or equivelent.

I think its important to support differentiation on composite structures.
This is the thing ChainRules.jl calls a `Composite` differential,
and Swift calls a Tangent Associated Type.

Reply:
Seems like this kinda thing works:
```
>=> foo1:: (2=>Real, 2=>Real)->Real
... foo1 (xs,ys) = sum for i. xs.i*ys.i

>=> foo ([1.0, 2.0], [3.0, 4.0])
Error: variable not in scope: foo
>=> foo1 ([1.0, 2.0], [3.0, 4.0])
11.0
>=> grad foo1
Type error: Can't print values of type: (((2=>Real), (2=>Real)) -> ((2=>Real), (2=>Real)))
>=> grad foo1 ([1.0, 2.0], [3.0, 4.0])
([3.0, 4.0], [1.0, 2.0])
```